src/ucs/memory/numa.c

```diff
@@ -72,3 +105,242 @@ int ucs_numa_node_of_cpu(int cpu)
+}
+
+int ucs_numa_num_configured_nodes()
+{
+    static int max_node = -1;
+
+    if (max_node == -1) {
+        UCS_NUMA_GET_MAX_DIRENT(max_node, UCS_NUMA_NODES_DIR_PATH, "node",
+                                UCS_NUMA_MAX_NODE, UCS_NUMA_DEFAULT_NODE);
+    }
+
+    return max_node;
+}
+
+int ucs_numa_num_configured_cpus()
+{
+    static int max_cpu = -1;
+
+    ucs_spin_lock(&ucs_numa_global_ctx.lock);
+    if (max_cpu == -1) {
+        UCS_NUMA_GET_MAX_DIRENT(max_cpu, UCS_NUMA_CORES_DIR_PATH, "cpu",
+                                (__CPU_SETSIZE - 1), 0);
+        max_cpu += 1;
+    }
+
+    ucs_spin_unlock(&ucs_numa_global_ctx.lock);
+    return max_cpu;
+}
+
+ucs_numa_node_t ucs_numa_node_of_cpu_v2(int cpu)
+{
+    /* Used for caching to improve perfromance */
+    static ucs_numa_node_t cpu_numa_node[__CPU_SETSIZE] = {0};
+    int node                                            = 0;
+    char core_dir_path[PATH_MAX];
+
+    ucs_assert(cpu < __CPU_SETSIZE);
+
+    ucs_spin_lock(&ucs_numa_global_ctx.lock);
+    if (cpu_numa_node[cpu] == 0) {
+        ucs_snprintf_safe(core_dir_path, PATH_MAX, UCS_NUMA_CORE_DIR_PATH, cpu);
+        UCS_NUMA_GET_MAX_DIRENT(node, core_dir_path, "node",
+                                ucs_numa_num_configured_nodes(),
+                                UCS_NUMA_DEFAULT_NODE);
+        cpu_numa_node[cpu] = node + 1;
+    }
+
+    ucs_spin_unlock(&ucs_numa_global_ctx.lock);
+    return cpu_numa_node[cpu] - 1;
+}
+
+ucs_numa_node_t ucs_numa_node_of_device(const char *dev_path)
+{
+    char dev_node_path[PATH_MAX];
+    FILE *dev_node;
+    int match;
+    int parsed_node;
+
+    ucs_snprintf_safe(dev_node_path, sizeof(dev_node_path), "%s/numa_node",
+                      dev_path);
+    dev_node = fopen(dev_node_path, "r");
+    if (!dev_node) {
+        ucs_debug("Failed to open sysfs node");
+        return UCS_NUMA_DEFAULT_NODE;
+    }
+
+    match = fscanf(dev_node, "%d\n", &parsed_node);
+    fclose(dev_node);
+    if (match <= 0) {
+        ucs_debug("Failed to parse sysfs node");
+        return UCS_NUMA_DEFAULT_NODE;
+    }
+
+    if ((parsed_node < 0) || (parsed_node > UCS_NUMA_MAX_NODE)) {
+        ucs_debug("Failed to parse sysfs node");
+        return UCS_NUMA_DEFAULT_NODE;
+    }
+
+    return parsed_node;
+}
+
+/**
+ * Parse and iterate the distance list of @param source.
+ * For each neighbor of source add distance(source, neighbor)
+ * to ucs_numa_global_ctx.numa_distance_hash.
+ *
+ * @param [in]  distances NUMA distances list string
+ * @param [in]  source    source NUMA node
+ * @param [in]  dest      destination NUMA node
+ *
+ * @return Distance from @param source to @param dest
+ */
+static ucs_numa_distance_t ucs_numa_node_put_distances(char *distances,
+                                                       ucs_numa_node_t source,
+                                                       ucs_numa_node_t dest)
+{
+    char *token = strtok(distances, UCS_NUMA_NODE_DISTANCE_DELIMITER);
```

Comment:
Maybe move `strtok` to a separate line.

Reply:
You mean something like this:
```
char *token = 
        strtok(distances, UCS_NUMA_NODE_DISTANCE_DELIMITER);
```
        
I can do that but according to clang it's fine.

Reply:
The second one.